File: cos382/A00_LanguageComparison/prolog/primepartitions.pl
isPrimeHelper(N, Divisor) :-
    N is Divisor -> true;
    0 is N mod Divisor -> false;
    isPrimeHelper(N, Divisor+1).

isPrime(N) :-
    N < 2 -> false; isPrimeHelper(N, 2).

primesHelper(Limit, PotentialPrime, PrimesList, Primes) :-
    Limit < 2 -> Primes = [];
    PotentialPrime > Limit -> Primes = PrimesList;
    isPrime(PotentialPrime) -> append(PrimesList, [PotentialPrime], NewPrimesList),
    Sum is PotentialPrime+1,
    primesHelper(Limit, Sum, NewPrimesList, Primes);
    Sum is PotentialPrime+1,
    primesHelper(Limit, Sum, PrimesList, Primes).

primes(Limit, Primes) :-
    primesHelper(Limit, 2, [], Primes).

primePartitionHelper(N, WorkingList, PrimesList, PrimePartition) :-
    N = 0 -> atomic_list_concat(WorkingList, " + ", Partition),
    write(Partition),
    nl;
    N < 0 -> PrimePartition = [];
    PrimesList = [] -> PrimePartition = [];
    PrimesList = [CPN|PrimesListTail],
    append(WorkingList, [CPN], NewWorkingList),
    Difference is N-CPN,
    primePartitionHelper(Difference, NewWorkingList, PrimesListTail, PrimePartition),
    primePartitionHelper(N, WorkingList, PrimesListTail, PrimePartition).

primePartition(PrimePartition) :-
    write("Choose number to partition: "),
    read(N),
    primes(N, Primes),
    primePartitionHelper(N, [], Primes, PrimePartition).

File: cos382/A00_LanguageComparison/haskell/primepartitions.hs
isPrimeHelper :: Integer -> Integer -> Bool
isPrimeHelper n divisor
    | divisor >= (n-1) = True
    | mod n divisor == 0 = False
    | otherwise = isPrimeHelper n (divisor+1)

isPrime :: Integer -> Bool
isPrime n
    | n < 2 = False
    | otherwise = isPrimeHelper n 2

primes :: Integer -> [Integer]
primes limit
    | limit < 2 = []
    | otherwise = filter (isPrime) [2..limit]

primePartitionListToString :: [Integer] -> [Char] -> [Char]
primePartitionListToString list workingList
    | list == [] = take (length(workingList)-3) $ workingList
    | otherwise = primePartitionListToString (tail list) (workingList ++ show(head list) ++ " + ")

primePartitionHelper :: Integer -> [Integer] -> [Integer] -> IO()
primePartitionHelper n workingList primesList =
    let cpn = head primesList
        primesListTail = tail primesList in
    if n == 0 then putStrLn(primePartitionListToString workingList "")
    else if n < 0 then return ()
    else if primesList == [] then return ()
    else
        primePartitionHelper (n-cpn) (workingList ++ [cpn]) (primesListTail) >>
        primePartitionHelper n workingList (primesListTail)

primePartition :: IO()
primePartition = do
    putStrLn "Choose number to partition"
    n <- getLine
    let intN = read n
    primePartitionHelper intN [] (primes intN)

main = primePartition
